pkg/ddevapp/config_merge.go

```diff
@@ -0,0 +1,206 @@
+package ddevapp
+
+import (
+	"errors"
+	"log"
+	"regexp"
+
+	"github.com/imdario/mergo"
+	"gopkg.in/yaml.v2"
+)
+
+// mergeConfigToApp does an unmarshall with merging
+func (app *DdevApp) mergeConfigToApp(source []byte) error {
+	// save away state before merges.
+	unmergedApp := &DdevApp{}
+	err := mergo.Merge(unmergedApp, app)
```

Comment:
If I'm not mistaken, all the logic you have here can be replaced with this pseudocode:

```
// start with app as the loaded official starting ddevapp
for _, i := range extraFiles {
        unmergedNewConfig := &DdevApp{}
	err = yaml.Unmarshal(i, unmergedNewConfig)
	if err != nil {
		return err
	}
        err = unmergedNewConfig.Validate()
         if err != nil {
		return err
	}
        err = mergo.Merge(app, unmergedNewConfig)
         if err != nil {
		return err
	}
}

What do you think about trying that out? It may require a little more work, but seems like mergo is doing the work it's intended to do here.

Reply:
Problem is that the app object wraps reference types like maps and slices.  Doing a copy as you suggest screwed up the reference types, with changes in app bleeding into unmergedApp. Hence the deep copy library.

Was a slightly puzzling issue that took a spell to debug.